Diver pattern for VanillaCloudFoundryApplication entity.

kiuby88 opened the pull request:
A `VanillaPaasApplicationCloudFoundryDriver` is a driver driver for  `VanillaCloudFoundryApplication` which is based on a interface namely `VanillaPaasApplicationDriver`.

The driver management is based on the currently Brooklyn driver pattern using the convention for the name driver interface name without the `Driver` suffix  + kind of location + `Driver` suffix (through `ReflectiveEntityDriverFactory`).

kiuby88 committed:
Diver pattern for VanillaCloudFoundryApplication entity.

Adding driver for start operation

kiuby88 committed:
Adding Stop operation to the driver

kiuby88 committed:
Adding set env operation to CloudFoundryPaasClient

Some tests were also added

kiuby88 committed:
Adding env management to Vanilla driver.

Some unit, live and yaml were added

kiuby88 commented:
@andreaturli what do you think?
Driver pattern looks interesting from my point of view

kiuby88 committed:
SetEnv effector to add env to applications

kiuby88 committed:
Deleting CLOUDFOUNDRY_LOCATION sensor

kiuby88 committed:
Litte issue in yml live test

kiuby88 committed:
Adding resources management to CloudFoundryPaasClient

kiuby88 committed:
Adding effectors to manage Memory, Disk and Instances

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaCloudfoundryApplication.java:
```diff
@@ -77,14 +82,38 @@
 
     AttributeSensor<Lifecycle> SERVICE_STATE_ACTUAL = Attributes.SERVICE_STATE_ACTUAL;
 
+    AttributeSensor<Map> APPLICATION_ENV = Sensors.builder(Map.class, "application.env").build();
```
What's the diff between ENV and APPLICATION_ENV?

kiuby88 replied:
`ENV` is a `ConfigKey` and `APPLICATION_ENV` is a sensor that shows the current env of an application. For example, if you add an env using the `setEnv` effector,  `APPLICATION_ENV` will show the new added env

andreaturli replied:
you may consider `AttributeSensorAndConfigKey`

kiuby88 replied:
Let me take a look. Thanks for the advice

kiuby88 replied:
Done. It was a good idea

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaCloudfoundryApplication.java:
```diff
@@ -77,14 +82,38 @@
+    AttributeSensor<Map> APPLICATION_ENV = Sensors.builder(Map.class, "application.env").build();
+
+    AttributeSensor<Integer> USED_INSTANCES =
+            Sensors.newIntegerSensor("cloudfoundry.application.instances",
```
Simply INSTANCES?

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaCloudfoundryApplication.java:
```diff
@@ -77,14 +82,38 @@
+    AttributeSensor<Integer> ALLOCATED_DISK =
+            Sensors.newIntegerSensor("cloudfoundry.application.disk", "Application allocated disk (MB)");
+
+    @Effector(description = "Set an environment variable that can be " + "retrieved by the web application")
```
Weird '+'

kiuby88 replied:
Oops

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaCloudfoundryApplication.java:
```diff
@@ -77,14 +82,38 @@
+    public void setEnv(@EffectorParam(name = "name", description = "Name of the variable") String name,
+                       @EffectorParam(name = "value", description = "Value of the environment variable") String value);
+
+    @Effector(description = "Set the instances number that will be user by the web application")
```
Set the desired number of instances...

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaCloudfoundryApplicationImpl.java:
```diff
@@ -101,6 +95,16 @@ private void initApplicationName() {
+    }
+
+    @Override
     protected void initEnrichers() {
```
Why initEnrichers needs to call of these lifecycle commands?

kiuby88 replied:
I found similar called in `SoftwareProcessImpl` ([here](https://github.com/apache/brooklyn-server/blob/0b9682b8830df7715bb8b579cfc40a1734f03038/software/base/src/main/java/org/apache/brooklyn/entity/software/base/SoftwareProcessImpl.java#L142)) and I decided to follow its example

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaCloudfoundryApplicationImpl.java:
```diff
@@ -381,16 +334,56 @@ public Boolean call() {
+    private VanillaPaasApplicationDriver doInitDriver(CloudFoundryPaasLocation location) {
+        if (driver != null) {
+            if ((driver instanceof VanillaPaasApplicationCloudFoundryDriver)
+                    && location.equals((driver).getLocation())) {
```
I think we should also check location is instanceOf PaasLocation or something

kiuby88 replied:
Actually, A `VanillaCloudFundryApplicationDriver#getLocation()` returns a `CloudFoundryPaasLocation` (which is a `PaasLocation`). so, may it be redundant?

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaPaasApplicationCloudFoundryDriver.java:
```diff
@@ -0,0 +1,193 @@
+
+
+    public VanillaPaasApplicationCloudFoundryDriver(VanillaCloudfoundryApplicationImpl entity,
+                                                    CloudFoundryPaasLocation location) {
```
Remove empty line

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaPaasApplicationCloudFoundryDriver.java:
```diff
@@ -0,0 +1,193 @@
+        //TODO
+    }
+
+
```
Remove empty line

andreaturli commented on src/test/java/org/apache/brooklyn/cloudfoundry/entity/VanillaCloudFoundryApplicationLiveTest.java:
```diff
@@ -58,7 +63,65 @@ public void testDeployApplicationWithoutDomain() throws IOException {
+
+        entity.setMemory(DOUBLE_MEMORY);
+        assertEquals(entity.getAttribute(VanillaCloudfoundryApplication.ALLOCATED_MEMORY).intValue(),
+                DOUBLE_MEMORY);
```
I think this constant doesnt make the test more readable :)

andreaturli replied:
I do like the driver pattern as well!

I think you are doing a great job! Keep up with the great work. 

Some minor comments but overall a +1

I'll run the tests asap though

kiuby88 replied:
I agree, I have opened a new task in trello in order to refactor the test. A lot of details can be addressing. I do not know how solving it, first I used numbers directly, after I used `MEMORY*2`, finally I decided to use contantes, probably I should use a different new as `NEW_MEMORY` or `NEW_MEMORY_AMOUNT`

kiuby88 replied:
Thanks a lot @andreaturli. 
In any case, I would like to refactor and simplify the test in order to avoid duplications.

kiuby88 committed:
Minor improvements
Empty lines
Improvement of some effector descriptions
Improvement of some constant names

kiuby88 committed:
Test refactorization

kiuby88 commented:
@andreaturli, I have refactored the test, I hope they will be more readable now :)

kiuby88 committed:
Adding a simple DownloaderResolver

kiuby88 committed:
Add restart effector

kiuby88 committed:
Adding host property to VanillaCloudFoundryApplication entity

kiuby88 committed:
Minor test refactoring

kiuby88 committed:
Typo: VanillaCloudFoundryApplication instead of VanillaCloudfoundryApplication

kiuby88 committed:
Split testModifyResourceProfile in different test which are focused on the different resources, memory, disk and instances

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaPaasApplicationDriver.java:
```diff
@@ -0,0 +1,46 @@
+
+    void delete();
+
+    void setEnv(Map<String, String> env);
```
not sure we want to expose those setters here

kiuby88 replied:
This setter is used by the vanilla's effector `setEnv` which allows to add more env to an application.
Do you think the aforementioned effector makes sense?

andreaturli replied:
isn't that an async operation? do you think it makes sense to consider something like

```
ProcessTaskWrapper<Integer> setEnv(Map<String, String> env);
```

?

kiuby88 replied:
I am not sure, I would say that I do not need that, but may I miss something

andreaturli replied:
ok let's keep it in mind

kiuby88 replied:
ok, thanks a lot

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaPaasApplicationCloudFoundryDriver.java:
```diff
@@ -0,0 +1,207 @@
+import com.google.common.collect.ImmutableList;
+
+public class
+        VanillaPaasApplicationCloudFoundryDriver implements VanillaPaasApplicationDriver {
```
weird indentation

kiuby88 replied:
+1

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaPaasApplicationCloudFoundryDriver.java:
```diff
@@ -0,0 +1,207 @@
+    private void postLaunch() {
+        entity.sensors().set(Attributes.MAIN_URI, URI.create(applicationUrl));
+        entity.sensors().set(VanillaCloudFoundryApplication.ROOT_URL, applicationUrl);
+        updateMemorySensor(entity.getConfig(VanillaCloudFoundryApplication.REQUIRED_MEMORY));
```
not sure I follow, doesn't CF expose a call to retrieve the status of the app? I think sensor values should come from that API call instead of being manually set. wdyt?

kiuby88 replied:
In fact, I was thinking about that. First, I used the API to set the value to the sensor. But after I decided to use the config instead, because the resources (memory, disk, and instances) value were just configured in the previous step (application creation). So, I thought the application configuration was very near and in `postlaunch` method we could assume that the application resource values in the cloud were the same that the config keys, so I did not need a API called.
But, I should not assume anything, so I agree with you, the sensor should be feed by the API (client).

kiuby88 replied:
Thanks for that suggestion @andreaturli 
I have modified that. Now the sensors are fed using the client.
I also updated all the necessary tests

kiuby88 committed:
Feeding resource sensors using the CloudFoundry API (client)

kiuby88 commented:
@andreaturli  I have addressed some suggestions, but I do not know why travis is not happy. I looks something about `brooklyn-downstream-parent-0.10.0-SNAPSHOT.pom` is wrong.

kiuby88 committed:
Deleting APPLICATION_ENV. ENV was refactorized in order to use BasicAttributeSensorAndConfigKey

andreaturli commented:
I think travis is happier now, repository.apache.org is behaving now.
Merging now! Great job, @kiuby88

kiuby88 commented:
Thanks a lot, @andreaturli
